x-pack/dockerlogbeat/pipelineManager/pipelineManager.go

```diff
@@ -0,0 +1,233 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package pipelinemanager
+
+import (
+	"crypto/sha1"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/idxmgmt"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/outputs"
+	"github.com/elastic/beats/libbeat/publisher/pipeline"
+	"github.com/elastic/beats/libbeat/publisher/processing"
+	"github.com/pkg/errors"
+	yaml "gopkg.in/yaml.v2"
+)
+
+// containerConfig is the common.Config unpacking type
+type containerConfig struct {
+	Pipeline pipeline.Config        `config:"pipeline"`
+	Output   common.ConfigNamespace `config:"output"`
+}
+
+// Pipeline represents a single pipeline and the count of associated clients
+type Pipeline struct {
+	pipeline *pipeline.Pipeline
+	refCount int
+}
+
+// PipelineManager is a handler into the map of pipelines used by the plugin
+type PipelineManager struct {
+	mu     *sync.Mutex
+	Logger *logp.Logger
+	// pipelines key: config hash
+	pipelines map[string]*Pipeline
+	// clients config: filepath
+	clients map[string]*ClientLogger
+}
+
+// NewPipelineManager creates a new Pipeline map
+func NewPipelineManager(logCfg *common.Config) *PipelineManager {
+	return &PipelineManager{
+		Logger:    logp.NewLogger("dockerlogbeat"),
+		mu:        new(sync.Mutex),
+		pipelines: make(map[string]*Pipeline),
+		clients:   make(map[string]*ClientLogger),
+	}
+}
+
+// CloseClientWithFile closes the client with the associated file
+func (pm *PipelineManager) CloseClientWithFile(file string) error {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
```

Comment:
We run 'CloseClientWithFile` in a separate go-routine (building an unbuffered 'channel'), because we don't want the API to block if CloseClientWithFile blocks. Due to the mutex here we do allow `CreateClientWithConfig` to block indirectly, though. Is this ok?

Reply:
Yes.  If `CreateClientWithConfig` bombs out while the HTTP handler is waiting for it, the container startup will fail and an error will be returned to the user via the CLI or whatever they're using to start. Otherwise a container might fail _after_ a user a started it, or we get into some weird scenario where a container is running but the user doesn't know it's not logging. I'm less worried about the container stop passing its errors to the user.

Reply:
To clarify: We are ok with `CreateClientWithConfig` and `startLoggingHandler` blocking?